plans/init.pp

```diff
@@ -0,0 +1,98 @@
+# Reboots targets and waits for them to be available again.
+#
+# @param nodes Targets to reboot.
+# @param message Message to log with the reboot (for platforms that support it).
+# @param reboot_delay How long (in seconds) to wait before rebooting. Defaults to 0, reboot immediately.
+#        Note on most Linux platforms, this time can only be specified in minutes, so non-0 values will
+#        delay reboot by at least 1 minute.
+# @param disconnect_wait How long (in seconds) to wait before checking whether the server has rebooted. Defaults to 1.
+# @param reconnect_wait How long (in seconds) to attempt to reconnect before giving up. Defaults to 120.
+# @param retry_interval How long (in seconds) to wait between retries. Defaults to 1.
+plan reboot (
+  TargetSpec $nodes,
+  Optional[String] $message = undef,
+  Integer[0] $reboot_delay = 0,
+  Integer[0] $disconnect_wait = 1,
+  Integer[0] $reconnect_wait = 120,
+  Integer[0] $retry_interval = 1,
+) {
+  $start_time = Timestamp()
+  $targets = get_targets($nodes)
+
+  # Get last boot time
+  $begin_boot_time_results = without_default_logging() || {
+    run_task('reboot::last_boot_time', $targets)
+  }
+
+  # Reboot; catch errors here because the connection may get cut out from underneath
+  $expect_disconnect = ($reboot_delay == 0)
+  $reboot_result = run_task('reboot', $nodes, timeout => $reboot_delay, message => $message, _catch_errors => $expect_disconnect)
+
+  # Wait long enough for all targets to trigger reboot, plus disconnect_wait to allow for shutdown time.
+  # Errored results should only occur if timeout was 0, so we assume they rebooted immediately.
+  $timeouts = $reboot_result.ok_set.map |$result| { $result['timeout'] }
+  $wait = max($timeouts + [0])
+  reboot::sleep($wait+$disconnect_wait)
```

Comment:
This is not a very efficient way of waiting. You might want to wait for 10min for the servers to go down because that's how long the prod ones take, but if you happen to target dev nodes that shut down in 30sec you don't want to wait the full 10min

Reply:
I lean towards not having a `disconnect_wait` at all and relying only on the checks for last boot time with `reconnect_wait`. That seems simpler to me. However if you know that shutting down takes a long time, waiting some period before checking would cause less network activity.

To make this work efficiently as wait for disconnect then wait for connect, sounds like we'd need two new methods in Bolt.

Reply:
I totally missed the whole uptime checking thing. Yes no need for a "disconnect wait" if we are checking uptime, but there does need to be a timeout of some kind so that if the server just decides it's not going to reboot we don't sit there waiting forever. I think the network activity would likely be an okay trade off for speed here, it's not likely to be much